File: Stiver_Problems/DP/P006.cpp
// P006.cpp - Move all zeroes to the end (stable)

#include <bits/stdc++.h>
using namespace std;

vector<int> brute(const vector<int>& nums) {
    int n = nums.size();
    vector<int> temp(n, 0);
    int k = 0;
    for (int i = 0; i < n; i++) {
        if (nums[i] != 0) {
            temp[k] = nums[i];
            k++;
        }
    }
    return temp;
}

vector<int> optimal(vector<int> nums) {
    int n = nums.size();
    int i = 0, j = 0;
    while (j < n) {
        if (nums[j] == 0) {
            j++;
            continue;
        }
        if (nums[i] == 0) {
            swap(nums[i], nums[j]);
            i++; j++;
        } else {
            i++; j++;
        }
    }
    return nums;
}

int main() {
    cout << "Enter elements: ";
    string line;
    getline(cin, line);
    stringstream ss(line);
    vector<int> nums;
    int x;
    while (ss >> x) nums.push_back(x);

    auto b = brute(nums);
    auto o = optimal(nums);

    cout << "Brute: ";
    for (int v : b) cout << v << " ";
    cout << "\nOptimal: ";
    for (int v : o) cout << v << " ";
    cout << "\n";
    return 0;
}

File: Arrays/P006.py
def brute(nums):
    n=len(nums)
    temp=[0]*n
    k=0
    for i in range(n):
        if nums[i]!=0:
            temp[k]=nums[i]
            k+=1
    return temp

def optimal(nums):
    n=len(nums)
    i=j=0
    while(j<n):
        if nums[j]==0:
            j+=1
            continue
        if nums[i]==0:
            nums[i],nums[j]=nums[j],nums[i]
            i+=1
            j+=1
    return nums

if __name__=='__main__':
    arr=input('Enter elements: ').split()
    nums=list(map(int,arr))
    print('Brute: ',brute(nums))
    print('Optimal: ',optimal(nums))